### DISCUSSION TITLE
Pass item in custom button controller

### DISCUSSION THREAD

**Discussion by dformdotdk** [joomengine/Joomla-Component-Builder#769] [Q&A] [answered]
I feel this is mostly me missing some basic knowledge, but I'm struggling with this. I have created a view called cancelsubscription where I've added a custom button cancelSubscription that targets the controller function below. I know I should not echo anything in a controller, but it's just for testing - later it will update the database.

```
public function cancelSubscription() {
  echo '<pre>'; print_r($this); echo '</pre>';
}
```

However, when I press the button, the view changes from view=cancelsubscription&id=1&Itemid=964 to &view=mysubscriptions&Itemid=964 and the id is lost. 

My question is: Why is the button creating this redirect when pressed, and how can I get the item object in the controller function?

I've seen this tutorial, but it didn't help me: https://www.youtube.com/watch?v=X6vyExgjMkU

**Comment by Llewellynvdm** [comment, marked as answer]
Try:
```php
public function cancelSubscription() {
  echo '<pre>'; print_r($this); echo '</pre>';exit;
}
```

**Reply by dformdotdk** [reply]
Tried, but makes no difference in terms of the redirect. And just to be clear - the button is added to a custom site view.

**Reply by Llewellynvdm** [reply]
The redirection is not yet set, and so if the button and the function lines-up your script should die, with the print out.

Dealing with the redirect you will need to add some more code, like this to the controller method:
```php
// Redirect to ....
$this->setRedirect(
	JRoute::_(
			'index.php?option=com_[[[component]]]......', false
	)
);
```

Then make sure the the ` Target Controller Method` name is the same as the function name `cancelSubscription`, also make sure the dynamicGet relation to the view is corresponding to your `Target`. This means if your main dynamicGet is a **list** return set, then target list, if it is a **single** return set then single.

Let me see.... hmmm that should be all need to get this to work, I use it all the time.

**Reply by dformdotdk** [reply]
Thank you - now I understand how to control the redirect. 

Getting the item id to the controller is another case. Looking at the $this object, I can see info about the view, itemid, task etc, but but I need to get the item id to the controller.

I've tried $id	= JFactory::getApplication()->input->getInt('id'); and other variations without any luck. 

Looking at the populateState function in the model, I see that the id is set:

```
// Get the itme main id
$id = $this->input->getInt('id', null);
$this->setState('cancelsubscription.id', $id);

```

Seems like I'm missing a link somewhere? - I really need to get a better understanding of MVC.

**Reply by dformdotdk** [reply]
I can add a hidden field with the id inside the view template, and retrieve than on submit, but that seems a bit "hackish" and not very elegant.

**Reply by dformdotdk** [reply]
Looking at the form action url for the view, the id is missing. If I manually override it in code, then I can get the id using `$this->input->getInt('id', null);` - so how can I affect the form action url and add the id to it? Seems it's all I need.

**Comment by dformdotdk** [comment]
Surely it's not the best solution, but a solution. I ended up writing to the Joomla session in my view.html.php file.

```
$session = JFactory::getSession();
$session->set('recordId', $this->item->id);
```

Then in my controller, I get the recordId using

```
$id = JFactory::getSession()->get('recordId');
```

I'm sure there is a better way, and something that is more elegant. Just don't know how. 
But I don't want to add hidden form fields containing the id - that's easy to tamper with. Even though I'll implement checks before updating the database, I don't like it.

**Reply by Llewellynvdm** [reply]
You should get the form post like this:
```php
// get the app object
$app = JFactory::getApplication();
$form = $this->input->get('jform', array(), 'ARRAY');
// check if the ID is not set and also not the GUID
if (!isset($form['id']))
{
	$form['id'] = $this->input->get('id', 0, 'INT');
}
```
